Term, Vote and RaftLog hygiene around shutdown and write calls

digitalstain opened the pull request:
Previous implementations did not properly check the return value of write() calls,
 so half written entries were a possibility. This ensures that the complete buffer
 contents always end up on disk or failure is raised.
Renames the previous Vote, Term and RaftLog tests to Contract tests to allow for
 file based, implementation specific testing.
Adds shutdown() lifecycle implementation to these three stores, to allow for closing
 of their file channels.
Code hygiene stuff - renames and javadocs

digitalstain committed:
Term, Vote and RaftLog file-based implementations use writeAll instead of write

Previous implementations did not properly check the return value of write() calls,
 so half written entries were a possibility. This ensures that the complete buffer
 contents always end up on disk or failure is raised.
Renames the previous Vote, Term and RaftLog tests to Contract tests to allow for
 file based, implementation specific testing.

digitalstain committed:
Slight reordering of operations, simply for readability purposes

digitalstain committed:
File based TermStore, VoteStore and RaftLog will now properly close their file channels

Adds shutdown() lifecycle method to DurableTermStore, DurableVoteStore and NaiveDurableRaftLog.
 They are introduced to the global lifecycle before any other raft component, so they are the last
 to be closed.

davidegrohmann commented on enterprise/core-edge/src/test/java/org/neo4j/coreedge/raft/log/NaiveDurableRaftLogContractTest.java:
```diff
@@ -0,0 +1,40 @@
+        fileSystem.mkdir( directory );
+
+        return new NaiveDurableRaftLog( fileSystem, directory, new DummyRaftableContentSerializer(), new Monitors() );
+    }
```
Having tests here describing the shutdown behaviour with suppressed exceptions would be great.

davidegrohmann replied:
Nevermind, I was looking in the wrong place.

digitalstain committed:
Do not interrupt threads running on a scheduled timer

Some of the threads running on a timer in core-edge will end up
 doing I/O ops when updating the instance state. This can, during
 shutdown, cause ClosedByInterruptException and leave the instance
 in an unknown state.
This change ensures that shutting down the scheduler service will allow
 threads to complete uninterruptibly.

digitalstain committed:
Renames TimeoutService to RenewableTimeoutService

digitalstain committed:
Renames ScheduledTimeout to ScheduledRenewableTimeout. Adds TimeoutService javadocs.

digitalstain committed:
Merge branch '3.0' into ce-store-hygiene

Conflicts:
	enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/replication/shipping/RaftLogShipper.java
	enterprise/core-edge/src/main/java/org/neo4j/coreedge/server/core/CoreServerStartupProcess.java
	enterprise/core-edge/src/test/java/org/neo4j/coreedge/CoreServerStartupProcessTest.java
